@font-face `font-family` descriptor should not allow a list of values

weinig opened the pull request:
#### 10a58951bcbd6b83478fd5798c22eaf8ec5e62dc
<pre>
@font-face `font-family` descriptor should not allow a list of values
<a href="https://bugs.webkit.org/show_bug.cgi?id=196381">https://bugs.webkit.org/show_bug.cgi?id=196381</a>
<a href="https://rdar.apple.com/142009630">rdar://142009630</a>

Reviewed by Darin Adler.

Updates @font-face&apos;s `font-family` descriptor consistently use a since CSSValue
rather than a CSSValueList with a single element as its representation.

Also requires piping a CSSValuePool in CSS::PropertyParserState so that worker
uses of parsing functions don&apos;t use the main thread only default pool.

* LayoutTests/fast/text/font-face-empty-string-expected.txt:
* LayoutTests/fast/text/font-face-empty-string.html:
* LayoutTests/fast/text/font-face-family-expected.txt:
* LayoutTests/fast/text/font-face-family.html:
    - Update tests/results to match the Font Loading spec which
      states that the normal grammar should be used for font-family.

* Source/WebCore/css/CSSFontFace.cpp:
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFaceSet.cpp:
* Source/WebCore/css/CSSFontFaceSet.h:
* Source/WebCore/css/CSSFontSelector.cpp:
* Source/WebCore/css/FontFace.cpp:
    - Switch to use single &quot;family&quot; rather than &quot;families&quot;.

* Source/WebCore/css/CSSProperties.json:
    - Switch to using a generated parser.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h:
(WebCore::CSSPropertyParserHelpers::parseFontFaceFontFamily):
(WebCore::CSSPropertyParserHelpers::consumeFontFaceFontFamily): Deleted.
    - Remove hand rolled consume function, but adds &quot;parse&quot; version
      to match other font-face descriptors for use by the FontFace class.

* Source/WebCore/svg/SVGFontFaceElement.cpp:
(WebCore::SVGFontFaceElement::attributeChanged):
    - Set a rule type on the CSSParserContext to make setProperty() call
      the right parsing function.

* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserState.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
    - Update to pass the appropriate CSSValuePool to the parser.

Canonical link: <a href="https://commits.webkit.org/293891@main">https://commits.webkit.org/293891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b515e993d781f16a1da6ae4277642f9d9f99e728

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19845 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105326 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50778 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28318 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76276 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90498 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56637 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15222 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8496 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50147 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107685 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27309 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20007 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85229 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27672 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86705 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84768 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21544 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29446 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7181 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21164 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27246 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32479 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27057 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30373 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28616 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -222,32 +219,30 @@ void CSSFontFaceSet::add(CSSFontFace& face)
-        String familyName = CSSFontFaceSet::familyNameFromPrimitive(downcast<CSSPrimitiveValue>(item));
-        if (familyName.isNull())
-            continue;
+    auto familyName = CSSFontFaceSet::familyNameFromPrimitive(downcast<CSSPrimitiveValue>(familyToSearchFor));
```
If this argument needs to be `CSSPrimitiveValue`, then why not use that type for the argument?

darinadler commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -169,30 +168,28 @@ String CSSFontFaceSet::familyNameFromPrimitive(const CSSPrimitiveValue& value)
-                ensureLocalFontFacesForFamilyRegistered(familyName);
-            familyFontFaces = { };
-        }
+    auto familyName = AtomString { CSSFontFaceSet::familyNameFromPrimitive(downcast<CSSPrimitiveValue>(*family)) };
```
What guarantees that `familyCSSValue` returns a `CSSPrimitiveValue`?

weinig replied:
Like the old code, nothing but the guarantees of the parser make this true. (I am sure people are sick of me saying it, but switching to the strong representations will fix this).

nt1m reviewed (commented):
The bug/commit title is slightly confusing. I understood it as "font-family descriptor allows a list of values so we're adding support for it", but looking at the commit more, that's not what this commit is doing

weinig committed:
@font-face `font-family` descriptor should not allow a list of values
https://bugs.webkit.org/show_bug.cgi?id=196381
rdar://142009630

Reviewed by Darin Adler.

Updates @font-face's `font-family` descriptor consistently use a since CSSValue
rather than a CSSValueList with a single element as its representation.

Also requires piping a CSSValuePool in CSS::PropertyParserState so that worker
uses of parsing functions don't use the main thread only default pool.

* LayoutTests/fast/text/font-face-empty-string-expected.txt:
* LayoutTests/fast/text/font-face-empty-string.html:
* LayoutTests/fast/text/font-face-family-expected.txt:
* LayoutTests/fast/text/font-face-family.html:
    - Update tests/results to match the Font Loading spec which
      states that the normal grammar should be used for font-family.

* Source/WebCore/css/CSSFontFace.cpp:
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFaceSet.cpp:
* Source/WebCore/css/CSSFontFaceSet.h:
* Source/WebCore/css/CSSFontSelector.cpp:
* Source/WebCore/css/FontFace.cpp:
    - Switch to use single "family" rather than "families".

* Source/WebCore/css/CSSProperties.json:
    - Switch to using a generated parser.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h:
(WebCore::CSSPropertyParserHelpers::parseFontFaceFontFamily):
(WebCore::CSSPropertyParserHelpers::consumeFontFaceFontFamily): Deleted.
    - Remove hand rolled consume function, but adds "parse" version
      to match other font-face descriptors for use by the FontFace class.

* Source/WebCore/svg/SVGFontFaceElement.cpp:
(WebCore::SVGFontFaceElement::attributeChanged):
    - Set a rule type on the CSSParserContext to make setProperty() call
      the right parsing function.

* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserState.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
    - Update to pass the appropriate CSSValuePool to the parser.

Canonical link: https://commits.webkit.org/293891@main
